### PR TITLE
feat: new js apis

### LINKED ISSUE #2233: [meta] Tracking issue for Node.js/Deno standard library inspired functions to include in `@tauri-apps/api`
I will be refactoring our js api and aiming to make the it more familiar to users coming from electron/js-land so it is easier to migrate. 

Some of current functions will be deprecated and removed when new variants are released.

We want to gather some feedback from users and which functions they want to add from the **Nodejs standard library**, so comment below and I will add them to this list of new functions:
1. `os` module
   - [x] `EOL`
   - [x] `platform()`
   - [x] `version()`
   - [x] `tempdir()`
   - [ ] `hostname`

2. `path` module
   - [x] `join()`
   - [x] `normalize()`
   - [x] `resolve()`
   - [x] `sep`
   - [x] `delimiter`
   - [x] `dirname()`
   - [x] `basename()`
   - [x] `extname()`
   - [x] `isAbsolute()`

3. `fs` module
	- [ ] `create()`
	- [ ] `open()`
    - [ ] `close()`
    - [ ] `FsFile` class
	- [ ] `writeFile()`
    - [ ] `writeTextFile()`
    - [ ] `readFile()`
    - [ ] `readTextFile()`
    - [ ] `copyFile()`
    - [ ] `readDir()`
    - [ ] `mkdir()`
    - [ ] `remove()`
    - [ ] `rename()`
    - [ ] `exists()`
 
4. `net` module
    - [ ] `createConnection()`
   
5. `dns` module

### LINKED ISSUE #4350: [bug] $platform-$arch id inconsistency
### Describe the bug

The platform id is inconsistent between the JS api and the launcher update json (and possibly elsewhere?)

For example as stated here: https://tauri.studio/v1/api/js/modules/os/  
the platform for windows, x86-64 should be `win32-x86_64`

But the platform id in the json of the tauri updater becomes `windows-x86_64`  
darwin/x86-64 also becomes an even weirder `darwin-intel`
https://docs.rs/tauri/1.0.0-rc.16/tauri/updater/index.html#update-file-json-format

Can we make this two consistent, to avoid confusion and manual translation between the two?

### Reproduction

_No response_

### Expected behavior

_No response_

### Platform and versions

```shell
1.0 rc16
```

### Stack trace

_No response_

### Additional context

_No response_

### LINKED ISSUE #4490: [feat] Universal Fetch API
### Describe the problem

Hey folks, I'm curious as to why Tauri does not opt-in for a kind of universal fetch polyfill (i.e. one that would call the browser `fetch` on the web, and Tauri's `http` module when running in Tauri). I'm attempting to build a hybrid application that I can deploy both to Tauri as well as a web browser (similar to how Notion's app supports both). I was surprised to learn that Tauri's `fetch` implementation is not compatible with the [function signature](https://github.com/microsoft/TypeScript/blob/main/lib/lib.dom.d.ts#L17373) of the browser `fetch`, resulting in me having to write a compatibility layer over it:

```typescript
export async function fetch(input: RequestInfo | URL, init?: RequestInit): Promise<Response> {
	const { fetch } = await import('@tauri-apps/api/http');

	const res = await fetch(input.toString(), {
		...init,
		method: init.method as HttpVerb,
		body: {
			type: 'Text',
			payload: init.body
		}
	});

	return new Response(JSON.stringify(res.data) as BodyInit, res);
}
```

Furthermore, I have to wrap calls to this fetch implementation in conditional statements that check if the app is running in Tauri. My app is in SvelteKit, and having to juggle calls to different `fetch` implementations from SvelteKit's `load` function, the browser client, and Tauri is all fairly cumbersome. 

### Describe the solution you'd like

As most of the issues I see on here regarding CORS and whatnot have the simple response of "just use Tauri's HTTP module", I think it could be beneficial to make the `fetch` API fully compatible with that of browsers, and automatically swap out the browser's `fetch` with Tauri's `fetch` when the app is built with the HTTP module enabled. 

I'm not sure if there's been a previous discussion around this, or if there were specific reasons the HTTP module has gone in the direction it did.

### Alternatives considered

At the very least, I think it could be good to make the `fetch` API compatible with that of browsers, even if a polyfill is not added automatically. This would prevent myself and other users from having to write shims to make TypeScript not complain. 

### Additional context

_No response_

### LINKED ISSUE #5432: [feat] http supports request progress and cancellation
### Describe the problem

Provides the way to capture request progress and cancel requests.

### Describe the solution you'd like

Similar to axios [progress capturing](https://github.com/axios/axios#-progress-capturing) and [cancellation](https://github.com/axios/axios#cancellation).

```js
import { http } from '@tauri-apps/api';

const controller = new AbortController();

const response = await http.fetch(downloadUrl, {
    method: 'GET',
    responseType: ResponseType.Binary,
    signal: controller.signal,
    onUploadProgress (e) {
        console.log(e.loaded);
        console.log(e.total);
    },
    onDownloadProgress (e) {
        console.log(e.loaded);
        console.log(e.total);
    }
});

controller.abort();
```

### Alternatives considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by amrbashir**
closes #2233 
closes #4490
closes #4350 

### What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [X] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?

- [X] Yes, and the changes were approved in issue #___
- [ ] No

### Checklist
- [X] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [X] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [X] I have added a convincing reason for adding this feature, if necessary

### Other information

Although some of the APIs intended in this PR could be add in a minor version of v1, The majority will be breaking changes so it is better to just include all in one PR for v2. I will also try to split the PR into a reviewable chunks.

## TODOS:
1. `os` module
   - [X] Remove `type()`
   - [X] Adjust `platform()` to match Rust's implementation.
   - [X] `hostname()`
   - [X] Rename `tempdir()` to `tempDir()`

2. `path` module
	- [X] `tempDir()`

2. `fs` module
	- [x] `create()`
	- [x] `open()`
    - [x] `close()`
    - [x] `FsFile` class
    	- [X] `close()`
    	- [X] `read()`
    	- [X] `write()`
    	- [x] `seek()`
    	- [x] `truncate()`
    	- [x] `stat()`
    - [x] `write()`
	- [X] `writeFile()`
    - [X] `writeTextFile()`
    - [X] Remove `writeBinaryFile()`
    - [x] `read()`
    - [X] `readFile()`
    - [X] `readTextFile()`
    - [X] `readTextFileLines()`
    - [X] Remove `readBinaryFile()`
    - [x] Adjust `copyFile()`
    - [x] Adjust `readDir()`
    - [x] `mkdir()`
    - [x] `remove()`
    - [x] `rename()`
    - [x] `seek()`
    - [x] `truncate()`
    - [x] `ftruncate()`
    - [x] `stat()`
    - [x] `lstat()`
    - [x] `fstat()`
    
3. `http` module
	- [x] Make `fetch` compatible with the web api
	  - [ ] stream `file://` on `reqwest-client` feature
	  - [ ] cancel `fetch` using `AbortController`
	- [x] Remove `Client` and `getClient`

4. `window` module
	- [x] Remove `appWindow`
	- [x] Combine `WindowManager`, `WebviewHandle` and `WebViewWindow` into one `Window` class.
    
- [ ] Merge all change files into one big change file

### CODE REVIEW

File: .changes/api-rs-fs-disk-entry.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+"api": "major"
+"tauri": "major"
+---
+
+**Breaking change** Removed `DiskEtry` struct from `tauri::api::dir`. Use `DirEntry` instead.
```
Comment (vdvman1) [inline .changes/api-rs-fs-disk-entry.md]: Typo: `DiskEtry` should be `DiskEntry`

Comment (amrbashir) [reply .changes/api-rs-fs-disk-entry.md]: thank you, I will keep this in mind when I merge all changelogs together

File: tooling/api/src/http.ts [hunk 2]
```diff
@@ -46,515 +46,76 @@
...
-  options?: FetchOptions
-): Promise<Response<T>> {
-  if (defaultClient === null) {
-    defaultClient = await getClient()
-  }
-  return defaultClient.request({
-    url,
-    method: options?.method ?? 'GET',
-    ...options
+  const res = new Response(Uint8Array.from(data), {
```
Comment (Yesterday17) [inline tooling/api/src/http.ts:110]: It seems not suitable for responses with large body size.

Comment (FabianLars) [reply tooling/api/src/http.ts:110]: The current implementation works the same, we currently can only send "normal" arrays to the backend, so unless the new IPC can support streaming bodies, this is what we get. :/

**Comment by jlarmstrongiv** [comment]
Are `fs` streaming APIs planned?  See https://github.com/tauri-apps/tauri/issues/2233#issuecomment-1510266483 and https://github.com/tauri-apps/tauri/issues/4133

**Comment by lucasfernog** [comment]
@amrbashir we'll need to find some time to move these to the plugins-workspace repo. I would wait a little longer on the FS APIs, since we'll see if [cap-std](https://github.com/bytecodealliance/cap-std) can help us there (see #6107).

**Comment by amrbashir** [comment]
sure, I will start on porting them to plugins-workspace this week

**Pull request closed without merging**